monitoring/aws.go

```diff
@@ -38,26 +43,81 @@ type awsHasProfileChecker struct{}
 
 // Name returns this checker name
 // Implements health.Checker
-func (*awsHasProfileChecker) Name() string {
+func (c *awsHasProfileChecker) Name() string {
 	return awsHasProfileCheckerID
 }
 
 // Check will check the metadata API to see if an IAM profile is assigned to the node
 // Implements health.Checker
-func (*awsHasProfileChecker) Check(ctx context.Context, reporter health.Reporter) {
-	session, err := session.NewSession()
+func (c *awsHasProfileChecker) Check(ctx context.Context, reporter health.Reporter) {
+	probes, err := c.check(ctx)
 	if err != nil {
-		reporter.Add(NewProbeFromErr(awsHasProfileCheckerID, "failed to create session", trace.Wrap(err)))
-		return
+		probes.Add(NewProbeFromErr(c.Name(), "failed to validate IAM profile", err))
+	}
+	health.AddFrom(reporter, probes)
+}
+
+// check will check the metadata API to see if an IAM profile is assigned to the node.
+func (c *awsHasProfileChecker) check(ctx context.Context) (probes health.Reporter, err error) {
+	probes = &health.Probes{}
+
+	sess, err := session.NewSession()
+	if err != nil {
+		return probes, trace.Wrap(err, "failed to create session")
+	}
+
+	metadata := ec2metadata.New(sess)
+	err = c.iamInfo(ctx, metadata)
+	if err != nil {
+		return probes, trace.Wrap(err, "failed to determine node IAM profile")
 	}
-	metadata := ec2metadata.New(session)
 
-	_, err = metadata.IAMInfo()
+	probes.Add(NewSuccessProbe(c.Name()))
+	return probes, nil
+}
+
+// iamInfo checks if an IAM profile is assigned to the node.
+// Takes a context for cancelation.
+//
+// This is a slight modification to `EC2Metadata.IAMInfo(...)`
+// https://github.com/aws/aws-sdk-go/blob/13ad2b09f6ff7e16808b4a45c32684e6c66bf86d/aws/ec2metadata/api.go#L92
+func (c *awsHasProfileChecker) iamInfo(ctx context.Context, metadata *ec2metadata.EC2Metadata) error {
```

Comment:
Hmm, is there a better way than copying code from the go-aws SDK? I'm looking at the docs and it seems like you can provide your own HTTP client in the config, have you tried that? https://docs.aws.amazon.com/sdk-for-go/api/aws/ec2metadata/#New

Reply:
I came up with the following (incl. the proper `context canceled` error message so we can revert the test):
```go
// check will check the metadata API to see if an IAM profile is assigned to the node.
func (c *awsHasProfileChecker) check(ctx context.Context) (probes health.Reporter, err error) {
	probes = &health.Probes{}

	sess, err := session.NewSession()
	if err != nil {
		return probes, trace.Wrap(err, "failed to create session")
	}

	config := sess.ClientConfig(ec2metadata.ServiceName)
	config.Config.HTTPClient = &http.Client{
		Transport: newTransport(ctx),
		Timeout:   5 * time.Second,
	}
	metadata := ec2metadata.NewClient(*config.Config, config.Handlers, config.Endpoint, config.SigningRegion)
	_, err = metadata.IAMInfo()
	if isContextCanceledError(err) {
		return probes, trace.Wrap(err, "context canceled")
	}
	if err != nil {
		return probes, trace.Wrap(err, "failed to determine node IAM profile")
	}

	probes.Add(NewSuccessProbe(c.Name()))
	return probes, nil
}

func newTransport(ctx context.Context) *transport {
	return &transport{
		ctx: ctx,
		t:   http.DefaultTransport,
	}
}

func (r *transport) RoundTrip(req *http.Request) (*http.Response, error) {
	req = req.WithContext(r.ctx)
	return r.t.RoundTrip(req)
}

type transport struct {
	ctx context.Context
	t   http.RoundTripper
}

func isContextCanceledError(err error) bool {
	log.WithError(err).Info("isContextCancelledError?")
	if err == nil {
		return false
	}
	switch err := unwrapAWSError(err).(type) {
	case *url.Error:
		log.WithError(err.Err).Info("url.Error.")
		return err.Err == context.Canceled
	default:
		log.WithError(err).Info("other error.")
		return strings.Contains(err.Error(), "context canceled")
	}
}

func unwrapAWSError(err error) error {
	for err != nil {
		switch origErr := err.(type) {
		case awserr.Error:
			err = origErr.OrigErr()
		default:
			return err
		}
	}
	return err
}
```